content/blog/withr-3-0-0/index.Rmd

```diff
@@ -0,0 +1,226 @@
+
+One benefit of the rewrite is that mixing withr tools and `on.exit()` in the same function now correctly interleaves cleanup:
+
+```{r}
+local({
+  on.exit(print(1))
+
+  withr::defer(print(2))
+
+  on.exit(print(3), add = TRUE, after = FALSE)
+
+  withr::defer(print(4))
+
+  print(5)
+})
+```
+
+But the main benefit is increased performance. Here is how `defer()` compared to `on.exit()` in the previous version:
+
+```{r, eval = FALSE}
+base <- function() on.exit(NULL)
+withr <- function() defer(NULL)
+
+# withr 2.5.2
+bench::mark(base(), withr(), check = FALSE)[1:8]
+#> # A tibble: 2 × 8
+#>   expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
+#>   <bch:expr> <bch:tm> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>
+#> 1 base()            0   82ns  6954952.        0B    696.   9999     1
+#> 2 withr()      26.2µs 27.9µs    35172.    88.4KB     52.8  9985    15
+```
+
+withr 3.0.0 has now caught up to `on.exit()` quite a bit:
+
+```{r, eval = FALSE}
+# withr 3.0.0
+bench::mark(base(), withr(), check = FALSE)[1:8]
+#> # A tibble: 2 × 8
+#>   expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
+#>   <bch:expr> <bch:tm> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>
+#> 1 base()            0   82ns  7329829.        0B       0  10000     0
+#> 2 withr()      2.95µs  3.4µs   280858.        0B     225.  9992     8
+```
+
+Of course `on.exit()` is still much faster, in part because `defer()` supports more features (more on that below), but mostly because `on.exit` is a primitive function whereas `defer()` is implemented as a normal R function. That said we hope that we now have made `defer()` (and the `local_` and `with_` functions that use it) sufficiently fast to be used even in performance-critical micro-tools.
+
+
+## Improved withr features
+
+Over the successive releases of withr we've improved the behaviour of cleanup expressions interactively, in scripts executed with `source()`, and in knitr. `on.exit()` is a bit inconsistent when it is used outside of a function:
+
+- Interactively, it doesn't do anything.
+- In `source()` and in knitr, it runs immediately instead of a the end of the script
+
+`withr::defer()` and the `withr::local_` helpers try to be more helpful for these cases.
+
+Interactively, it saves the cleanup action in a special global hook and you get information about how to actually perform the cleanup:
+
+```{r, eval = FALSE}
+file <- withr::local_tempfile()
+#> Setting global deferred event(s).
+#> i These will be run:
+#>   * Automatically, when the R session ends.
+#>   * On demand, if you call `withr::deferred_run()`.
+#> i Use `withr::deferred_clear()` to clear them without executing.
+
+# Clean up now
+withr::deferred_run()
+#> Ran 1/1 deferred expressions
+```
+
+In knitr or `source()`[^1], the cleanup is performed at the end of the document or of the script. If you need chunk-wide cleanup, use `local()` as we've been doing in the examples of this blog post:
```

Comment:
```suggestion
In knitr or `source()`[^1], the cleanup is performed at the end of the document or of the script. If you need chunk-level cleanup, use `local()` as we've been doing in the examples of this blog post:
```

Reply:
I wonder if folks are going to ask for some nicer syntax for this, i.e. something you can do in the chunk options to eliminate the visible `local()` call in the code.

Reply:
It also occurs to me that `local()` isn't very appealing in knitr because you'll no longer get input code and output results interleaved correctly.